### PR TITLE
Feature/misc improvements

### LINKED ISSUE #185: Add option to only generate observations for satellites that require retasking
Save some compute on multiagent sims.

### LINKED ISSUE #187: Add minimum period for communication evaluation
Prevent overheads to due communication module being triggered very frequently when many satellites are present

### LINKED ISSUE #188: Cache `agents` list
Gets called a lot on a given step, and should change during that time.

### LINKED ISSUE #189: Allow `ImagingSatellite` to have a different default request type for filtering than target

### LINKED ISSUE #191: `Eclipse` observation is too slow
**Describe the bug**
Slows down the execution of the sim pretty poorly.

**To Reproduce**
Compare sat with and w/o Eclipse in observation.

### PR DESCRIPTION
**Pull request by Mark2000**
## Description
Closes #189 
Closes #185 
Closes #188 
Closes #187
Closes #191 

Misc improvements. Review by commit.

### Type of change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### How should this pull request be reviewed?
- [X] By commit
- [ ] All changes at once

## How Has This Been Tested?

Tests updated.

### Passes Tests
- [X] __Unit tests__ `pytest --cov bsk_rl --cov-report term-missing tests/unittest`
- [X] __Integrated tests__ `pytest --cov bsk_rl --cov-report term-missing tests/integration`
- [X] __Documentation builds__ `cd docs; make html`

### Test Configuration
- Python: 3.10
- Basilisk: 2.4.17
- Platform: MacOS

# Checklist:

- [X] My code follows the style guidelines of this project (passes Black, ruff, and isort)
- [X] I have performed a self-review of my code
- [X] I have commented my code in hard-to-understand areas
- [X] I have made corresponding changes to the documentation and release notes
- [X] Commit messages are atomic, are in the form `Issue #XXX: Message` and have a useful message
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] If I changed an example ipynb, I have locally rebuilt the documentation

### CODE REVIEW

File: src/bsk_rl/gym.py [hunk 1]
```diff
@@ -507,17 +507,27 @@ def reset(
...
-        ]
+        if (
+            self._agents_last_compute_time is None
+            or self._agents_last_compute_time != self.simulator.sim_time
+        ):
+            truncated = super()._get_truncated()
+            agents = [
+                satellite.name
+                for satellite in self.satellites
+                if (satellite.is_alive() and not truncated)
```
Comment (LorenzzoQM) [inline src/bsk_rl/gym.py:542]: Truncated is just one boolean for the episode and not for each satellite, right? I guess you could just check if it is truncated before iterating over the satellites to make it faster? Still, I am not sure if this has any practical advantages.

File: src/bsk_rl/gym.py [hunk 2]
```diff
@@ -306,7 +311,18 @@ def _get_obs(self) -> MultiSatObs:
         Returns:
             tuple: Joint observation
         """
-        return tuple(satellite.get_obs() for satellite in self.satellites)
+        if self.generate_obs_retasking_only:
```
Comment (LorenzzoQM) [inline src/bsk_rl/gym.py:316]: Nice!

File: src/bsk_rl/gym.py (hunk 1 above)
Comment (Mark2000) [reply src/bsk_rl/gym.py:542]: I think this was copying some example from petting zoo. I don’t see that being a huge performance difference.

File: docs/source/release_notes.rst [hunk 3]
```diff
@@ -19,6 +19,7 @@ Development Version
   performance.
 * Add option to ``generate_obs_retasking_only`` to prevent computing observations for
   satellites that are continuing their current action.
+* Improve performance of :class:`~bsk_rl.obs.Eclipse` observations by about 95%.
```
Comment (LorenzzoQM) [inline docs/source/release_notes.rst:26]: Excellent change!

Comment (LorenzzoQM) [reply docs/source/release_notes.rst:26]: Do we need to add the change in the access filters here?

Comment (Mark2000) [reply docs/source/release_notes.rst:26]: Sure, I'll note that it's possible to restrict access filters to only act on a certain type of opportunity. The base filter behavior should be the same.

**Comment by Mark2000** [comment]
Added a docstring for a method that was missing one

**Review by LorenzzoQM — APPROVED** [review]
Good changes and excellent performance improvement!

**Pull request merged**